### PR TITLE
Fix scaling of glyphs and improve type safety in the MCU backend

### PR DESCRIPTION
**Pull request by tronical**
The code was mixing logical and physical sizes, causing glyphs being
doubly scaled down. Instead, this patch introduces:

 * Physical* and Logical* euclid length/size/rect aliases
 * some extraction traits for getting the scalars in rects/sizes as lengths (until euclid has them
built-in)
 * wrapper traits/types for safely extracting the physical font metrics the
 compiler generates (i16)
 * Fix a bug in the text height calculation where we failed to take the
   descent into account

### CODE REVIEW

File: internal/backends/mcu/lengths.rs [hunk 1]
```diff
@@ -0,0 +1,87 @@
+// Copyright © SixtyFPS GmbH <info@slint-ui.com>
```
Comment (tronical) [inline internal/backends/mcu/lengths.rs:1]: This feels like something that could be in the core library, so that we could have `geometry()` on the `Item` trait for example return a `LogicalRect`. Or the compiler could generate `PhysicalLength` for embedded glyph metrics. The other backends would probably benefit as well.

The main issue I see is that PhysicalLength should be f32 for GL/Qt and i16 for the mcu.

File: internal/backends/mcu/renderer.rs [hunk 2]
```diff
@@ -369,21 +372,27 @@ impl PrepareScene {
...
+                                .round_in()
+                                .cast(),
                         )
                     }) {
                         let actual_x = dest_rect.origin.x - t.rect.origin.x;
                         let actual_y = dest_rect.origin.y - t.rect.origin.y;
                         let stride = t.rect.width() as u16 * bpp(t.format);
                         self.new_scene_texture(
-                            dest_rect.cast().scale(sx, sy),
+                            LogicalRect::from_untyped(&dest_rect.cast().scale(sx, sy)),
```
Comment (tronical) [inline internal/backends/mcu/renderer.rs:388]: It would be really nice to get rid of this call.

Generally, the intersection with the clip rect is unfortunate. A few lines above we call `to_untyped()` (want to avoid) and here we call `from_untyped` again. For the glyphs we start out in a physical space but have to "temporarily" move to logical space for the interaction and then go back to physical for the `new_scene_texture` call (well, going back is done inside).

File: internal/backends/mcu/fonts.rs [hunk 3]
```diff
@@ -7,14 +7,66 @@ use core::cell::RefCell;
...
 
 thread_local! {
     static FONTS: RefCell<Vec<&'static BitmapFont>> = RefCell::default()
 }
 
-pub const DEFAULT_FONT_SIZE: f32 = 12.;
+#[derive(Copy, Clone)]
+pub struct Glyph(&'static BitmapGlyph);
+
+impl Glyph {
```
Comment (ogoffart) [inline internal/backends/mcu/fonts.rs:24]: Ideally BitmapGlyph would be using PhysicalRect or something like that.

File: internal/backends/mcu/lengths.rs [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+pub struct LogicalPx;
+pub type LogicalLength = euclid::Length<f32, LogicalPx>;
+pub type LogicalRect = euclid::Rect<f32, LogicalPx>;
+pub type LogicalPoint = euclid::Point2D<f32, LogicalPx>;
+pub type LogicalSize = euclid::Size2D<f32, LogicalPx>;
+
+pub type ScaleFactor = euclid::Scale<f32, LogicalPx, PhysicalPx>;
+
+pub trait SizeLengths<LengthType> {
+    fn width_length(&self) -> LengthType;
```
Comment (ogoffart) [inline internal/backends/mcu/lengths.rs:19]: the euclid PR uses get_width and get_height  for this kind of stuff.
Did you depart from it on purpose?

File: internal/backends/mcu/renderer.rs [hunk 5]
```diff
@@ -203,41 +207,32 @@ impl Scene {
...
 #[derive(Clone, Copy)]
 struct SceneItem {
-    x: u16,
-    y: u16,
-    width: u16,
-    height: u16,
+    x: PhysicalLength,
+    y: PhysicalLength,
+    width: PhysicalLength,
+    height: PhysicalLength,
```
Comment (ogoffart) [inline internal/backends/mcu/renderer.rs:227]: Candidate for a Rect

File: internal/backends/mcu/renderer.rs [hunk 6]
```diff
@@ -270,15 +265,15 @@ struct SceneTexture {
     format: PixelFormat,
     /// bytes between two lines in the source
     stride: u16,
-    source_width: u16,
-    source_height: u16,
+    source_width: PhysicalLength,
+    source_height: PhysicalLength,
```
Comment (ogoffart) [inline internal/backends/mcu/renderer.rs:269]: candidate for a size

File: internal/backends/mcu/renderer.rs [hunk 7]
```diff
@@ -304,45 +299,53 @@ struct PrepareScene {
...
             textures: vec![],
             state_stack: vec![],
             current_state: RenderState {
                 alpha: 1.,
-                offset: PointF::default(),
-                clip: RectF::new(PointF::default(), size / scale_factor.0),
+                offset: LogicalPoint::default(),
+                clip: LogicalRect::new(
+                    LogicalPoint::default(),
+                    size.cast() * scale_factor.inverse(),
```
Comment (ogoffart) [inline internal/backends/mcu/renderer.rs:313]: so size / scale_factor  didn't work?

File: internal/backends/mcu/renderer.rs (hunk 2 above)
Comment (ogoffart) [reply internal/backends/mcu/renderer.rs:388]: Also the image size is maybe not strictly speaking physical pixel, but texture pixel or something.
But i don't think that euclid has a type to scale these which is not the same for width and height anyway.
Also we go back to and from logical a bit too much there.  I wonder if it wouldn't be nicer if new_scene_texture takes a pixel in physicial pixel.

**Review by ogoffart — APPROVED** [review]

File: internal/backends/mcu/renderer.rs (hunk 2 above)
Comment (tronical) [reply internal/backends/mcu/renderer.rs:388]: Yes, I was thinking the same about `new_scene_texture`. I'll give it a try and see if it simplifies the code.

File: internal/backends/mcu/fonts.rs (hunk 3 above)
Comment (tronical) [reply internal/backends/mcu/fonts.rs:24]: Yes. That requires the compiler to emit that and the lengths types to be in i_slint_core. That's weird because of the PhysicalLength using i16 as scalar type.

File: internal/backends/mcu/renderer.rs (hunk 7 above)
Comment (tronical) [reply internal/backends/mcu/renderer.rs:313]: Hm, it didn't work earlier. But let me try again, I think I may have tried without the `.cast()`.

**Comment by tronical** [comment]
I'll merge this to master, rebase wip/mcu and try the rect/size/new_scene_texture in a separate change.

**Comment by tronical** [comment]
Landed in master as commit 02dcf1173bf98bbf439704eee3e1c048b5bd3617
